### PR TITLE
Update schedulers guide

### PR DESCRIPTION
**Pull request by hanlint**
Fixes the guides portion of #398 .

This PR:
* updates the guides on schedulers, and expands the section on SSR
* replaces several blocks with `.. testcode::`
* codeblocks that run `from composer import Trainer` previously overrode the patches in `doctest_fixtures`. Now, in the fixtures we directly monkeypatch `composer.Trainer`.

### CODE REVIEW

File: docs/source/doctest_fixtures.py [hunk 1]
```diff
@@ -100,3 +107,6 @@
     train_dataloader=train_dataloader,
     eval_dataloader=eval_dataloader,
 )
+
+# patch composer so that 'from composer import Trainer' calls do not override change above
+composer.Trainer = Trainer
```
Comment (hanlint) [inline docs/source/doctest_fixtures.py:112]:
```suggestion
composer.Trainer = Trainer

```

File: docs/source/trainer/using_the_trainer.rst [hunk 2]
```diff
@@ -193,31 +195,51 @@ Optimizers & Schedulers
...
+    trainer = Trainer(model=model,
+                      train_dataloader=train_dataloader,
+                      max_duration='90ep',
+                      schedulers=MultiStepScheduler(
+                          milestones=['0.3dur', '0.5dur', '0.9dur'],
+                          gamma=0.1
+                      ))
+
+
+See :doc:`Schedulers`<schedulers>` for details.
```
Comment (jbloxham) [inline docs/source/trainer/using_the_trainer.rst:242]:
```suggestion
See :doc:`Schedulers<schedulers>` for details.
```

File: docs/source/trainer/using_the_trainer.rst [hunk 3]
```diff
@@ -193,31 +195,51 @@ Optimizers & Schedulers
 ~~~~~~~~~~~~~~~~~~~~~~~
 
 You can easily specify which optimizer and learning rate scheduler to
-use during training. Composer provides a library of various optimizers
-and schedulers, but you can also include one of your own.
+use during training. Composer supports both pytorch schedulers as
```
Comment (jbloxham) [inline docs/source/trainer/using_the_trainer.rst:198]: Nit:

```suggestion
use during training. Composer supports both PyTorch schedulers as
```

File: composer/trainer/trainer.py [hunk 4]
```diff
@@ -179,27 +179,28 @@ class Trainer:
...
+
+                To see the difference, consider training for half as long using a cosine
+                annealing learning rate schedule. If the schedule is not rescaled,
+                training ends while the learning rate is still ~0.5 of the initial LR.
+                If the schedule is rescaled with ``scale_schedule_ratio``, the LR schedule
+                would finish the entire cosine curve, ending with a learning rate near zero.
+
         step_schedulers_every_batch (bool, optional): By default, native
             `PyTorch schedulers <https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate>`_
             are updated every epoch, while :mod:`composer schedulers<composer.optim>` are updated every step.
```
Comment (jbloxham) [inline composer/trainer/trainer.py:200]: This syntax causes "composer schedulers" to be rendered in monospace font, as if "composer schedulers" itself is the name of the module. It would be ideal to just link to the relevant page.

File: docs/source/trainer/schedulers.rst [hunk 5]
```diff
@@ -76,12 +74,60 @@ Below are the supported schedulers found at :mod:`composer.optim.scheduler`.
     CosineAnnealingWarmRestartsScheduler
     PolynomialScheduler
 
+.. note::
+
+    Compared to pytorch schedulers, :class:`.ComposerScheduler` need not be provided
```
Comment (jbloxham) [inline docs/source/trainer/schedulers.rst:79]:
```suggestion
    Compared to PyTorch schedulers, :class:`.ComposerScheduler` need not be provided
```

File: docs/source/trainer/schedulers.rst [hunk 6]
```diff
@@ -76,12 +74,60 @@ Below are the supported schedulers found at :mod:`composer.optim.scheduler`.
...
+    an ``optimizer`` directly. The trainer will handle binding the optimizer when
+    converting our schedulers to :class:`~torch.optim.lr_scheduler.LabmdaLR`.
```
Comment (jbloxham) [inline docs/source/trainer/schedulers.rst:81]: I think this is a little more detail than needs to be given. We can just say something like "The training will attach the optimizer later when it compiles the scheduler."

File: docs/source/trainer/schedulers.rst [hunk 7]
```diff
@@ -76,12 +74,60 @@ Below are the supported schedulers found at :mod:`composer.optim.scheduler`.
     CosineAnnealingWarmRestartsScheduler
     PolynomialScheduler
 
+.. note::
+
+    Compared to pytorch schedulers, :class:`.ComposerScheduler` need not be provided
+    an ``optimizer`` directly. The trainer will handle binding the optimizer when
```
Comment (jbloxham) [inline docs/source/trainer/schedulers.rst:80]: Also,  "optimizer" probably doesn't need to monospaced here.

**Review by jbloxham — APPROVED** [review]
Looks good to me! Just some copy-editing suggestions.

File: composer/trainer/trainer.py (hunk 4 above)
Comment (hanlint) [inline composer/trainer/trainer.py:200]:
```suggestion
            are updated every epoch, while :doc:`Composer Schedulers</trainer/schedulers>` are updated every step.
```

File: docs/source/trainer/schedulers.rst (hunk 7 above)
Comment (hanlint) [inline docs/source/trainer/schedulers.rst:80]:
```suggestion
    an optimizer directly. The trainer will handle binding the optimizer when
```

File: docs/source/trainer/schedulers.rst (hunk 6 above)
Comment (hanlint) [inline docs/source/trainer/schedulers.rst:81]:
```suggestion
    it compiles the scheduler later.
```
